### PR TITLE
Fixed #719: fix generate for union methods.

### LINKED ISSUE #719: External URL ref and oneOf broken generated code
### Version
master (SHA fec990c8f82)

### Config
```
package: openapi
generate:
  gorilla-server: true
  client: true
  models: true
output: pkg/openapi/generated.go
import-mapping:
  'https://raw.githubusercontent.com/deepmap/oapi-codegen/master/examples/petstore-expanded/petstore-expanded.yaml': github.com/deepmap/oapi-codegen/examples/petstore-expanded
```

### Spec 
```
openapi: "3.0.0"
info:
  version: 1.0.0
  title: Example broken generation with oneOf and external remote ref
paths:
  /check:
    get:
      summary: Return broken example
      responses:
        '200':
          description: Ok
          content:
            application/json:
              schema:
                items:
                  $ref: '#/components/schemas/exampleSchema'
components:
  schemas:
    exampleSchema:
      properties:
        item:
          type: object
          oneOf:
            - $ref: 'https://raw.githubusercontent.com/deepmap/oapi-codegen/master/examples/petstore-expanded/petstore-expanded.yaml#/components/schemas/NewPet'
      required:
        - item
```

#### Running command `oapi-codegen --config config.yaml spec.yaml` failed with error 

Error `openapi.go:61:51: expected '(', found '.' (and 2 more errors)`

If watch output, we'll see broken generated methods `As/From/Merge` for `exampleSchema.item`: generator put union type name as is with dot in name `externalRef\d+.SchemaName`: 
```
// AsexternalRef0.NewPet returns the union data inside the ExampleSchema_Item as a externalRef0.NewPet
        func (t ExampleSchema_Item) AsexternalRef0.NewPet() (externalRef0.NewPet, error) {
            var body externalRef0.NewPet
            err := json.Unmarshal(t.union, &body)
            return body, err
        }

        // FromexternalRef0.NewPet overwrites any union data inside the ExampleSchema_Item as the provided externalRef0.NewPet
        func (t *ExampleSchema_Item) FromexternalRef0.NewPet (v externalRef0.NewPet) error {
            b, err := json.Marshal(v)
            t.union = b
            return err
        }

        // MergeexternalRef0.NewPet performs a merge with any union data inside the ExampleSchema_Item, using the provided externalRef0.NewPet
        func (t *ExampleSchema_Item) MergeexternalRef0.NewPet (v externalRef0.NewPet) error {
            b, err := json.Marshal(v)
            if err != nil {
              return err
            }

            merged, err := runtime.JsonMerge(b, t.union)
            t.union = merged
            return err
        }
```

#### Generated code with enabled gorilla-server does not generate import gorilla/mux

This just can be fixed by adding 
```
additional-imports:
  - package: github.com/gorilla/mux
```

But... seems like it must be automatically added by enabling gorilla generator.

### PR DESCRIPTION
**Pull request by insidieux**
Fix #719

### CODE REVIEW

**Comment by insidieux** [comment]
Any updates? May be something wrong in PR? It will be cool to merge and release, cause at this moment we're using built binary from fork...

**Comment by masu-mi** [comment]
I'm running into this problem on my private project.

I was looking into submitting a PR and I found this.
I tried this changes on my own and confirmed the problem went away. I'm going to use replace_of directive until merge this PR. Thank you.

**Comment by jamietanna** [comment]
Hey, sorry we've not gotten around to reviewing just yet - will try to in the coming week.

Mind running `make generate` and committing the generated files?

**Comment by insidieux** [comment]
> Hey, sorry we've not gotten around to reviewing just yet - will try to in the coming week.
> 
> Mind running `make generate` and committing the generated files?

Already done. Nothing new was generated, not changes.

**Comment by insidieux** [comment]
@jamietanna Hi there! =) Any updates? Can we merge it?

File: pkg/codegen/templates/imports.tmpl [hunk 1]
```diff
@@ -26,7 +27,8 @@ import (
 	"github.com/getkin/kin-openapi/openapi3"
 	"github.com/go-chi/chi/v5"
 	"github.com/labstack/echo/v4"
-	"github.com/gin-gonic/gin"
+	"github.com/gin-gonic/gin"{{ if $globalOpts.Generate.GorillaServer }}
+	"github.com/gorilla/mux"{{ end }}
```
Comment (jamietanna) [inline pkg/codegen/templates/imports.tmpl:31]: were these changes required for this change to work?

**Review by jamietanna — COMMENTED** [review]
I'm having trouble picturing this in use - it'd be great to get an example spec added that uses this under the hood, so we have some test data that can then help us avoid breaking changes in the future - would that be possible? :pray:

File: pkg/codegen/templates/imports.tmpl (hunk 1 above)
Comment (deepmap-marcinr) [reply pkg/codegen/templates/imports.tmpl:31]: The code formatting removes unused imports, so we don't usually add them conditionally.

Comment (insidieux) [reply pkg/codegen/templates/imports.tmpl:31]: Did i understand correctly, what i just need to add `github.com/gorilla/mux` import and remove condition?

**Comment by insidieux** [comment]
> I'm having trouble picturing this in use - it'd be great to get an example spec added that uses this under the hood, so we have some test data that can then help us avoid breaking changes in the future - would that be possible? 🙏

Example is in linked issue 
I'll add it here 
```yaml
openapi: "3.0.0"
info:
  version: 1.0.0
  title: Example broken generation with oneOf and external remote ref
paths:
  /check:
    get:
      summary: Return broken example
      responses:
        '200':
          description: Ok
          content:
            application/json:
              schema:
                items:
                  $ref: '#/components/schemas/exampleSchema'
components:
  schemas:
    exampleSchema:
      properties:
        item:
          type: object
          oneOf:
            - $ref: 'https://raw.githubusercontent.com/deepmap/oapi-codegen/master/examples/petstore-expanded/petstore-expanded.yaml#/components/schemas/NewPet'
      required:
        - item
```

File: pkg/codegen/templates/imports.tmpl (hunk 1 above)
Comment (jamietanna) [reply pkg/codegen/templates/imports.tmpl:31]: Yes exactly, just removing the conditional should work.

That being said, I'm not sure if these changes were needed for this PR to work?

Comment (insidieux) [reply pkg/codegen/templates/imports.tmpl:31]: I can move it to another PR) 
As i said in Issue, without this fix (adding import), users must add additional import to config file.

Comment (insidieux) [reply pkg/codegen/templates/imports.tmpl]: I made a change: remove global var and remove conditional import.

**Comment by insidieux** [comment]
"Somewhere in the world a kitten is sad at the moment" (с)
But, if seriously, is there any updates?

**Comment by deepmap-marcinr** [comment]
I never thought when I wrote the original code that we'd be fetching specs over the internet, because that causes a version tracking issue, so we've always cached them locally and updated periodically. As far as oapi-codegen goes, this change looks pretty simple, but your example schema should be included in our tests, to ensure we don't regress on this particular issue in the future.

**Comment by insidieux** [comment]
> I never thought when I wrote the original code that we'd be fetching specs over the internet, because that causes a version tracking issue, so we've always cached them locally and updated periodically. As far as oapi-codegen goes, this change looks pretty simple, but your example schema should be included in our tests, to ensure we don't regress on this particular issue in the future.

Ok. Thanks for reply.
I'll add new tests for this case and 'll return to you)

**Comment by insidieux** [comment]
@deepmap-marcinr, hi there! 
I've updated PR: added test-case yaml file and added test for this case.

I'll be grateful if you check this out and merge if everything is ok)

**Comment by insidieux** [comment]
Guys! Any updates? 2 months delaying PR... =(

**Comment by insidieux** [comment]
Guys, it's me again =)
Let's review and merge? =)

**Comment by deepmap-marcinr** [comment]
Done, looks good, thanks.

________________________________________
